Adding geolocation corrections (#341)
infinum/eightshift-libs: src/Geolocation/Readme.md

Files used:
* manifest.json

# WP-ROCKET Plugin usage

> When geolocation is used in combination with WP-Rocket plugin it will not work by default because the cookie is set after the page is loaded and at that point it is too late to provide content.

Reviewer: ```suggestion
When geolocation is used in combination with the WP-Rocket plugin it will not work by default because the cookie is set after the page is loaded and at that point, it is too late to provide alternative content.
```

> WP-Rocket plugin provides options/hooks to fix this. By using hook `rocket_advanced_cache_file` we are able to inject our custom function in the page generation process inside the `wp-content/advanced-cache.php` file. By modifying this file we can detect users geolocation from the IP and set the cookie manually before the page is loaded. This way WP-Rocket can detect that cookie and provide the necessary cache. Below is the process of setting it:

Reviewer: ```suggestion
WP-Rocket plugin provides options/hooks to fix this. By using the `rocket_advanced_cache_file` hook we are able to inject our custom function into the page generation process inside the `wp-content/advanced-cache.php` file. By modifying this file we can detect the users' geolocation from the IP and set the cookie manually before the page is loaded. This way WP-Rocket can detect that cookie and provide the necessary cache. Below is the process of setting it:
```

> WP-Rocket plugin provides the options/hooks to fix this. By using this hook `rocket_advanced_cache_file` we are able to inject our custom function in the page generation process inside the `wp-content/advanced-cache.php` file. By modifying this file we can detect users geolocation from the IP and set the cookie manually before the page is loaded. This way WP-Rocket can detect that cookie and provide the necessary cache. Here is the process of setting it:

Reviewer 2: ```suggestion
WP-Rocket plugin provides options/hooks to fix this. By using hook `rocket_advanced_cache_file` we are able to inject our custom function in the page generation process inside the `wp-content/advanced-cache.php` file. By modifying this file we can detect users geolocation from the IP and set the cookie manually before the page is loaded. This way WP-Rocket can detect that cookie and provide the necessary cache. Below is the process of setting it:
```

## Set dynamic cookie:

> You must provide the list of cookies that will be used as a dynamic cookie to generate cached versions.

Reviewer 2: ```suggestion
You must provide the list of cookies that will be used as dynamic cookies to generate cached versions.
```

**Filter:**
```php
\add_filter('rocket_cache_dynamic_cookies', [$this, 'dynamicCookiesList']);
```

**Callback:**
```php
/**
 * List all dynamic cookies that will create new cached version.
 *
 * @param array<string, mixed> $items Items from the admin.
 *
 * @return array<int|string, mixed>
 */
public function dynamicCookiesList(array $items): array
{
	$items[] = 'esForms-country';

	return $items;
}
```

## Set custom function to advanced-cached.php file.

> By providing custom custom to advanced-cached.php file you will be able to detect geolocation and set cookies before the cached version is provided.

Reviewer 3: ```suggestion
By providing a custom function in the advanced-cached.php file, you will be able to detect geolocation and set cookies before the cached version is provided.
```

> By providing custom custom to advanced-cached.php file you will be able to detect geolocation and set cookies before the cached version is provided.

Reviewer 2: ```suggestion
By providing custom function to advanced-cached.php file you will be able to detect geolocation and set cookies before the cached version is provided.
```

> By providing a custom function in the advanced-cached.php file, you will be able to detect geolocation and set cookies before the cached version is provided.

Reviewer: ```suggestion
By providing a custom function in the `advanced-cached.php` file, you will be able to detect geolocation and set cookies before the cached version is provided.
```

> For `$esFormsPath` variable you must provide the absolute path to the `geolocationDetect.php` file in vendor.

Reviewer: ```suggestion
For the `$esFormsPath` variable you must provide the absolute path to the `geolocationDetect.php` file located in the `vendor` folder.
```

The provided example works if you have Eightshift-forms in your project.

**Filter:**
```php
\add_filter('rocket_advanced_cache_file', [$this, 'addNginxAdvanceCacheRules']);
```

**Callback:**
```php

> /**
>  * Add geolocation function in advance-cache.php config file on plugin activation used only with Nginx.
>  *
>  * @param string $content Original file output.
>  */

Reviewer: ```suggestion
/**
 * Add geolocation function in the advance-cache.php config file on plugin activation
 *
 * Used only with Nginx web servers.
 *
 * @param string $content Original file output.
 */
```

public function addNginxAdvanceCacheRules( string $content ) : string {
	$position = \strpos($content, '$rocket_config_class');

	// This part is string on purpose.
	$cors_function = '
	$esFormsPath = ABSPATH . "wp-content/plugins/eightshift-forms/src/Geolocation/geolocationDetect.php";
	if (file_exists($esFormsPath)) {
		require_once $esFormsPath;
	};';

	return \substr_replace($content, $cors_function, $position, 0);
}
```

## Disable `setLocationCookie` filter in your project.

> If you have installed Geolocation.php class in your project and you have this filter set, make sure to disable it, or provide the check to be disabled when WP-Rocket cache plugin is active.

Reviewer: ```suggestion
If you have installed Geolocation.php class in your project and you have this filter set, make sure to disable it, or provide the check to be disabled when the WP-Rocket cache plugin is active.
```

> If you have installed Geolocation.php class in your project and you have this filter set, make sure you disable it, or provide the check to be disabled when WP-Rocket cache plugin is active.

Reviewer 2: ```suggestion
If you have installed Geolocation.php class in your project and you have this filter set, make sure to disable it, or provide the check to be disabled when WP-Rocket cache plugin is active.
```
